### ISSUE TITLE
media-gfx/freecad-0.18.4-r1: build failure using boost-1.73.0

### ISSUE THREAD

**Issue by waebbl** [waebbl/waebbl-gentoo#236]
Build fails with boost-1.73.0 with instantiation errors.

```
FAILED: src/Mod/Path/App/CMakeFiles/Path.dir/Area.cpp.o 
/usr/bin/mpicxx -DHAVE_CONFIG_H -DHAVE_LIMITS_H -DPath_EXPORTS -DQT_CORE_LIB -DQT_NO_DEBUG -DQT_XML_LIB -D_OCC64 -Dqh_QHpointer -DvtkRenderingContext2D_AUTOINIT="1(vtkRenderingContextOpenGL2)" -DvtkRenderingCore_AUTOINIT="3(vtkInteractionStyle,vtkRenderingFreeType,vtkRenderingOpenGL2)" -I. -I/usr/include/vtk-8.2 -I/usr/include/eigen3 -I/usr/include/freetype2 -I/usr/include/double-conversion -Isrc -I/var/tmp/portage/media-gfx/freecad-0.18.4-r1/work/FreeCAD-0.18.4/src -Isrc/Mod/Path/App -I/usr/lib64/opencascade-7.4.0/ros/include/opencascade -I/usr/include/python3.7m -isystem /usr/include/qt5 -isystem /usr/include/qt5/QtCore -isystem /usr/lib64/qt5/mkspecs/linux-g++ -isystem /usr/include/qt5/QtXml  -Wall -Wextra -Wno-write-strings -O2 -pipe -march=bdver2 -fstack-protector-strong -fstack-check -g -std=c++11  -D_OCC64 -fPIC   -fext-numeric-literals -pthread -fPIC -MD -MT src/Mod/Path/App/CMakeFiles/Path.dir/Area.cpp.o -MF src/Mod/Path/App/CMakeFiles/Path.dir/Area.cpp.o.d -o src/Mod/Path/App/CMakeFiles/Path.dir/Area.cpp.o -c /var/tmp/portage/media-gfx/freecad-0.18.4-r1/work/FreeCAD-0.18.4/src/Mod/Path/App/Area.cpp
In file included from /usr/include/boost/geometry.hpp:17,
                 from /var/tmp/portage/media-gfx/freecad-0.18.4-r1/work/FreeCAD-0.18.4/src/Mod/Path/App/Area.cpp:35:
/usr/include/boost/geometry/geometry.hpp:27:138: note: #pragma message: CAUTION: Boost.Geometry in Boost 1.73 deprecates support for C++03 and will require C++14 from Boost 1.75 onwards.
 BOOST_PRAGMA_MESSAGE("CAUTION: Boost.Geometry in Boost 1.73 deprecates support for C++03 and will require C++14 from Boost 1.75 onwards.")
                                                                                                                                          ^
/usr/include/boost/geometry/geometry.hpp:28:110: note: #pragma message: CAUTION: Define BOOST_GEOMETRY_DISABLE_DEPRECATED_03_WARNING to suppress this message.
 BOOST_PRAGMA_MESSAGE("CAUTION: Define BOOST_GEOMETRY_DISABLE_DEPRECATED_03_WARNING to suppress this message.")
                                                                                                              ^
In file included from /usr/include/boost/geometry/index/detail/rtree/node/node.hpp:30,
                 from /usr/include/boost/geometry/index/rtree.hpp:60,
                 from /usr/include/boost/geometry/algorithms/detail/within/multi_point.hpp:33,
                 from /usr/include/boost/geometry/algorithms/detail/within/implementation.hpp:42,
                 from /usr/include/boost/geometry/algorithms/detail/covered_by/implementation.hpp:25,
                 from /usr/include/boost/geometry/algorithms/covered_by.hpp:24,
                 from /usr/include/boost/geometry/algorithms/detail/distance/multipoint_to_geometry.hpp:22,
                 from /usr/include/boost/geometry/algorithms/detail/distance/implementation.hpp:26,
                 from /usr/include/boost/geometry/algorithms/detail/comparable_distance/implementation.hpp:22,
                 from /usr/include/boost/geometry/algorithms/comparable_distance.hpp:23,
                 from /usr/include/boost/geometry/algorithms/detail/buffer/buffered_piece_collection.hpp:28,
                 from /usr/include/boost/geometry/algorithms/detail/buffer/buffer_inserter.hpp:35,
                 from /usr/include/boost/geometry/algorithms/buffer.hpp:41,
                 from /usr/include/boost/geometry/geometry.hpp:63,
                 from /usr/include/boost/geometry.hpp:17,
                 from /var/tmp/portage/media-gfx/freecad-0.18.4-r1/work/FreeCAD-0.18.4/src/Mod/Path/App/Area.cpp:35:
/usr/include/boost/geometry/index/detail/rtree/node/variant_visitor.hpp: In instantiation of ‘void boost::geometry::index::detail::rtree::apply_visitor(Visitor&, boost::variant<boost::geometry::index::detail::rtree::variant_leaf<Value, Parameters, Box, Allocators, Tag>, boost::geometry::index::detail::rtree::variant_internal_node<Value, Parameters, Box, Allocators, Tag> >&) [with Visitor = boost::geometry::index::detail::rtree::visitors::insert<WireJoiner::VertexInfo, boost::geometry::index::rtree<WireJoiner::VertexInfo, boost::geometry::index::linear<16>, WireJoiner::PntGetter>::members_holder, boost::geometry::index::detail::rtree::insert_default_tag>; Value = WireJoiner::VertexInfo; Parameters = boost::geometry::index::linear<16>; Box = boost::geometry::model::box<boost::geometry::model::point<double, 3, boost::geometry::cs::cartesian> >; Allocators = boost::geometry::index::detail::rtree::allocators<boost::container::new_allocator<WireJoiner::VertexInfo>, WireJoiner::VertexInfo, boost::geometry::index::linear<16>, boost::geometry::model::box<boost::geometry::model::point<double, 3, boost::geometry::cs::cartesian> >, boost::geometry::index::detail::rtree::node_variant_static_tag>; Tag = boost::geometry::index::detail::rtree::node_variant_static_tag]’:
/usr/include/boost/geometry/index/rtree.hpp:1545:37:   required from ‘void boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::raw_insert(const value_type&) [with Value = WireJoiner::VertexInfo; Parameters = boost::geometry::index::linear<16>; IndexableGetter = WireJoiner::PntGetter; EqualTo = boost::geometry::index::equal_to<WireJoiner::VertexInfo>; Allocator = boost::container::new_allocator<WireJoiner::VertexInfo>; boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::value_type = WireJoiner::VertexInfo]’
/usr/include/boost/geometry/index/rtree.hpp:688:15:   required from ‘void boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::insert(const value_type&) [with Value = WireJoiner::VertexInfo; Parameters = boost::geometry::index::linear<16>; IndexableGetter = WireJoiner::PntGetter; EqualTo = boost::geometry::index::equal_to<WireJoiner::VertexInfo>; Allocator = boost::container::new_allocator<WireJoiner::VertexInfo>; boost::geometry::index::rtree<Value, Options, IndexableGetter, EqualTo, Allocator>::value_type = WireJoiner::VertexInfo]’
/var/tmp/portage/media-gfx/freecad-0.18.4-r1/work/FreeCAD-0.18.4/src/Mod/Path/App/Area.cpp:670:40:   required from here
/usr/include/boost/geometry/index/detail/rtree/node/variant_visitor.hpp:51:25: error: no matching function for call to ‘apply_visitor(boost::geometry::index::detail::rtree::visitors::insert<WireJoiner::VertexInfo, boost::geometry::index::rtree<WireJoiner::VertexInfo, boost::geometry::index::linear<16>, WireJoiner::PntGetter>::members_holder, boost::geometry::index::detail::rtree::insert_default_tag>&, boost::variant<boost::geometry::index::detail::rtree::variant_leaf<WireJoiner::VertexInfo, boost::geometry::index::linear<16>, boost::geometry::model::box<boost::geometry::model::point<double, 3, boost::geometry::cs::cartesian> >, boost::geometry::index::detail::rtree::allocators<boost::container::new_allocator<WireJoiner::VertexInfo>, WireJoiner::VertexInfo, boost::geometry::index::linear<16>, boost::geometry::model::box<boost::geometry::model::point<double, 3, boost::geometry::cs::cartesian> >, boost::geometry::index::detail::rtree::node_variant_static_tag>, boost::geometry::index::detail::rtree::node_variant_static_tag>, boost::geometry::index::detail::rtree::variant_internal_node<WireJoiner::VertexInfo, boost::geometry::index::linear<16, 4>, boost::geometry::model::box<boost::geometry::model::point<double, 3, boost::geometry::cs::cartesian> >, boost::geometry::index::detail::rtree::allocators<boost::container::new_allocator<WireJoiner::VertexInfo>, WireJoiner::VertexInfo, boost::geometry::index::linear<16, 4>, boost::geometry::model::box<boost::geometry::model::point<double, 3, boost::geometry::cs::cartesian> >, boost::geometry::index::detail::rtree::node_variant_static_tag>, boost::geometry::index::detail::rtree::node_variant_static_tag> >&)’
     boost::apply_visitor(v, n);
     ~~~~~~~~~~~~~~~~~~~~^~~~~~
In file included from /usr/include/boost/variant/apply_visitor.hpp:16,
                 from /usr/include/boost/geometry/algorithms/append.hpp:26,
                 from /usr/include/boost/geometry/algorithms/detail/assign_values.hpp:33,
                 from /usr/include/boost/geometry/algorithms/detail/assign_indexed_point.hpp:21,
                 from /usr/include/boost/geometry/policies/relate/intersection_points.hpp:20,
                 from /usr/include/boost/geometry/strategies/intersection_strategies.hpp:20,
                 from /usr/include/boost/geometry/strategies/strategies.hpp:39,
                 from /usr/include/boost/geometry/geometry.hpp:58,
                 from /usr/include/boost/geometry.hpp:17,
                 from /var/tmp/portage/media-gfx/freecad-0.18.4-r1/work/FreeCAD-0.18.4/src/Mod/Path/App/Area.cpp:35:
/usr/include/boost/variant/detail/apply_visitor_unary.hpp:46:1: note: candidate: ‘template<class Visitor, class Visitable> typename Visitor::result_type boost::apply_visitor(Visitor&, Visitable&&)’
 apply_visitor(Visitor& visitor, Visitable&& visitable)
 ^~~~~~~~~~~~~
/usr/include/boost/variant/detail/apply_visitor_unary.hpp:46:1: note:   template argument deduction/substitution failed:
```

**Comment by waebbl** [comment]
[freecad-0.18.4-r1:20200518-114618.log](https://github.com/waebbl/waebbl-gentoo/files/4644201/freecad-0.18.4-r1.20200518-114618.log)

**Comment by granquet** [comment]
I'm getting the same issue.
long story short, after rebuilding boost 1.73, the problem is gone and I was able to build freecad (albeit 9999, not 18.4) with boost-1.73

long story longer... (spent 3hours on this one...)
you would need this boost patch: https://github.com/boostorg/geometry/pull/722/files

which is scheduled for boost 1.74.
So I tried to hack a boost 1.74 rc1 ebuild... but it was too much work for my liking.
In the meantime, the patch mentionned above is already pulled by gentoo upstream boost ebuild.
see /usr/portage/dev-libs/boost/files/boost-1.73-geometry-private-inheritance.patch

so I just rebuilt boost 1.73, rebuilt freecad and everything was fine.
don't know what happened on my system, maybe one overlay had pulled some boost 1.73 without the patch ? I would have expected the gentoo people to make a 1.73-r1 or something if the patch was missing at first...

easy to check if you are suffering from the same issue:
does /usr/include/boost/geometry/index/detail/rtree/visitors/insert.hpp at line 268 declares the visitor base class public?

if not, make sure you rebuild the official gentoo boot 1.73 ebuild, which should get you sorted out...
